Blocks dynamically become text blocks upon call to editableRef

nathggns opened the pull request:
## 🌟 What is the purpose of this PR?

This is a blocker for our implementation of the [Hook Service](https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0430-hook-service.md)

The hook service does not inform an embedding application that a block makes use of it statically. This is a change from our `editableRef` implementation, which appears in the block schema. This means we need to remove all places where we check a block's schema to work out whether to treat it as a text block.

This means we can't set a component type as a text block or not text block in the Prosemirror schema, but have to enable a block to become editable at any point in the future. This is a big shift from how it works right now. 

This touched some of the deepest fundamentals of how our Prosemirror / collab / entity store / blockprotocol / HASH UI layers are implemented / connected and there was lots of cruft / duplication / unstable and difficult to reason about bits, so this PR has had to change a lot of that in order to work out the current behaviour / change it. 

<!-- Explain, at a high level, what this does and why. -->
<!-- Use the 'What does this change?' section to list more specific implementation details. -->

## 🔗 Related links

<!-- Add links to any context it is worth capturing (e.g. Issues, Discussions, Discord, Asana) -->
<!-- Mark any links which are not publicly accessible as _(internal)_ -->
<!-- Don't rely on links to explain the PR, especially internal ones: use the sections above -->

- [Asana task](https://app.asana.com/0/1201095311341924/1202593774648017/f) _(internal)_
- [Hook service implementation plan](https://www.notion.so/hashintel/Hook-Service-implementation-replacing-editableRef-a899a8d86cd642fcaabd0119775bc94b) _(internal)_

## 🚫 Blocked by

<!-- If the pull request is blocked by anything, list the blockers here. -->
<!-- If applicable, link to them. -->

N/A

## 🔍 What does this change?

<!-- Use a bullet list to explain your changes in more detail, if it would be helpful. -->
<!-- If applicable, link to the specific commit.-->

- Use existing `ProsemirrorSchemaManager#ensureBlocksDefined` when receiving steps in collab instead of reimplementing
- `BlockConfig` has been renamed to `HashBlockMeta` and `BlockMeta` has been renamed to `HashBlock` to better reflect what they are, and the properties on the latter have been shortened too. The `Block` type has been removed. `fetchBlockMeta` has also been renamed to `fetchBlock` and generally uses of `config` and `meta` have been clarified/tidied up. This is cached in local storage, so the local storage key has been changed to prevent crashes
- New function `prepareBlockCache` to prefill the client side cache with block data passed from server to remove need to request it again 
- `isSwappable` has been renamed to `isTextBlock`, to better reflect what it is, and moved into `@hashintel/hash-shared` for use in the entity store/Prosemirror setup. In the long run we'll want to remove this, but that's for a later PR. `isTextBlock` is used sparingly and is essentially used to work out if we want to keep text content when switching between *our* text blocks. External blocks using `editableRef` don't get to keep the text data – this will rely on data mapping to fix. 
- Schema management code has been moved out of `ProsemirrorSchemaManager` which has now been renamed to `ProsemirrorManager` and has been almost entirely rewritten for reduction of code duplication / clarity. 
- Support adding text entity links to entities on update, not just on creation (in `save.ts` and `updatePageContents.ts`) – this is as we've moved to text entities being linked when the component first calls `editableRef`, rather than ahead of time
- The block context menu now works before the block has been saved, as it now uses `draftId` instead of `entityId` – there was no need to use the latter
- When `editableRef` is called, we now check if there is a text entity linked to the block, and create one if not. This is moving code from the entity store plugin/ProsemirrorSchemaManager, which used to do this ahead of time. This is because we no longer check ahead of time because we don't know if the block is a text block until it calls `editableRef`. We also check if it was *just* inserted by the suggester plugin (which has a new piece of state / actions) to know whether we need to update the cursor position to focus inside the new block, as the suggester plugin can't do that ahead of time either. This required adding a new `onBlockLoaded` prop to `BlockLoader`, so we could clear the "suggested" state if `editableRef` isn't immediately called (you don't want a block to be able to grab focus by calling `editableRef` at some undefined point in the future). To enable this, *all* blocks are now registered as "text blocks" in the Prosemirror schema, and we dynamically control editability using `contentEditable` properties – this is because it does not appear possible to "upgrade" a non-text block to a text block in the Prosemirror schema. Our usual schema mutations trick was not sufficient. This means we can no longer use `isTextBlock` field on Prosemirror nodes, as it will always be true.
- Remove duplication in defining blocks in prosemirror schema when loading a page from collab
-  Some document editing code has been moved from the suggester plugin into the new `ProsemirrorManager` 
- `useRemoteBlock` has been rewritten for clarity and to enable loading a block outside of a react component. I ended up not needed this, but no need to revert it
- `createDraftIdForEntity` is now `generateDraftIdForEntity` and now always generates random draft IDs – this will enable future work to support multiple copies of the same block on the same page and prevents its improper use
- Listeners have been moved out of the entity store plugin state so they can be updated without a transaction (i.e, mid transaction). This improves stability 

## 📜 Does this require a change to the docs?

<!-- If this adds a user facing feature or modifies how an existing feature is used, it likely needs a docs change. -->

- The docs for x need updating to explain that y

## ⚠️ Known issues

<!-- Are there known issues / intentionally omitted functionality? Flag them here to save reviewers doing so -->

N/A

## 🐾 Next steps

<!-- Are there are planned/suggested follow ups which are related but won't be done in this PR? -->

- Remove legacy text links
- Allow for text entities to be linked on any property, not just "text" properties
- Pass the plain text version of text as requested to blocks  - https://github.com/hashintel/hash/pull/883
- Replace editableRef with the hook service
- Allow multiple editable islands 

## 🛡 What tests cover this?

<!-- What automated tests cover this? Existing ones? New ones? None? -->

- Playwright tests

## ❓ How to test this?

<!-- Tell reviewers how they can test the functionality -->

This is a fairly fundamental rewrite of some of the deepest layers of HASH. You should do a full UAT of all known features.

## 📹 Demo

There is no visible change in this PR.

nathggns commented on packages/hash/frontend/src/blocks/page/ComponentView.tsx:
```diff
@@ -196,20 +204,127 @@ export class ComponentView implements NodeView<Schema> {
+      tr ??= state.tr;
+
+      const newTextDraftId = generateDraftIdForEntity(null);
+      addEntityStoreAction(state, tr, {
+        type: "newDraftEntity",
+        payload: {
+          accountId: childTextEntity.accountId,
+          draftId: newTextDraftId,
+          entityId: null,
+        },
+      });
+
+      // @todo should we use the text entity directly, or just copy the content?
+      addEntityStoreAction(state, tr, {
+        type: "updateEntityProperties",
+        payload: {
+          draftId: newTextDraftId,
+          // @todo indicate the entity type?
+          properties: textBlockNodeToEntityProperties(this.node),
+          merge: false,
+        },
+      });
+
+      addEntityStoreAction(state, tr, {
+        type: "updateEntityProperties",
+        payload: {
+          draftId: childTextEntity.draftId,
+          properties: {
+            text: {
+              __linkedData: {},
+              data: entityStorePluginStateFromTransaction(tr, state).store
+                .draft[newTextDraftId]!,
+            },
+          },
+          merge: true,
+        },
+      });
```
This is just moved from the block creation point to the point at which the block becomes editable

nathggns commented on packages/hash/shared/src/ProsemirrorSchemaManager.ts:
```diff
@@ -455,20 +293,21 @@ export class ProsemirrorSchemaManager {
-  // @todo handle saving the results of this
-  // @todo handle non-intermediary entities
-  async createRemoteBlockTr(
+  /**
+   * @todo This probably creates entities where it could/should reuse existing entities
+   */
+  async createBlock(
```
Need to investigate this

teenoh commented on packages/hash/api/src/collab/save.ts:
```diff
@@ -80,10 +82,8 @@ const ensureEntityTypeForComponent = async (
-      JSON.stringify(componentMeta.componentSchema),
-    );
+    const block = await fetchBlock(componentId);
+    const jsonSchema = JSON.parse(JSON.stringify(block.schema));
```
I understand this was not introduced in this PR, but why do we stringify and parse here?

nathggns replied:
It's a crude way of cloning it, because I mutate it below to remove `editableRef` from the schema if its there in order to ensure it doesn't end up in the relevant entity type.

teenoh commented on packages/hash/shared/src/entityStorePlugin.ts:
```diff
@@ -123,6 +123,8 @@ export const entityStorePluginStateFromTransaction = (
  * Do NOT change the entity's draftId mid-session - leave it as fake. If you
  * need to recall the entity's draftId, use mustGetDraftEntityForEntityId
+ *
+ * @todo should maybe always be random to handle a block being present twice
```
A bit curious, In what scenario will an instance of block entity be present twice (same entityId). I know it's possible for a block to show up multiple times (same entityTypeId, different entityId's), but can't think of a scenario where we'd have 2 blocks with the same entityId

nathggns replied:
It's long been an ambition in HASH to support having the same block on multiple pages, or the same block multiple times on one page (it's the "shared block" idea). It's possible we end up implementing this with multiple block entities pointing to the same data entity, but that wasn't how it was done when this was supported in HASH in the past. 

This is an out of date diff though. This has been changed now to always return a random ID.

CiaranMn commented on packages/hash/frontend/src/blocks/page/ComponentView.tsx:
```diff
@@ -196,20 +204,124 @@ export class ComponentView implements NodeView<Schema> {
+  }
+
+  private onBlockLoaded = () => {
+    /**
+     * After two calls to setImmediate, we know the block will have had the
+     * chance to initiate an editable section and thereby become selected if
+     * initiated via the suggester. If it hasn't happened in that time, we want
+     * to expire the opportunity to become automatically selected.
+     *
+     * @todo find a better way of knowing there's been the opportunity for
+     *       this which doesn't depend on timing
+     */
+    if (this.wasSuggested) {
```
This may relate to my `wasSuggested` comment above, but I think we could usefully explain what 'automatically selected' is somewhere.

nathggns replied:
Could you check if this is clearer (combined with my new comment on `wasSuggested`)?

[`e1db5b3` (#872)](https://github.com/hashintel/hash/pull/872/commits/e1db5b3c8515bb32f68c89a44f1004d6e6175912)

CiaranMn replied:
It is, thank you

CiaranMn commented on packages/hash/frontend/src/blocks/page/createSuggester/createSuggester.tsx:
```diff
@@ -121,35 +137,57 @@ export const createSuggester = (
-            return { ...state, disabled: true };
+            return { ...state, disabled: true, suggestedBlockPosition: null };
+
+          case "key":
+            return { ...state, suggestedBlockPosition: null };
+
+          case "suggestedBlock":
+            return {
+              ...state,
+              suggestedBlockPosition: action.payload.position,
+            };
         }
 
+        const suggestedBlockPosition =
+          state.suggestedBlockPosition === null ||
+          tr.selectionSet ||
+          docChangedInTransaction(tr)
+            ? null
+            : tr.mapping.map(state.suggestedBlockPosition);
```
Can we add a comment explaining this please – either inline or in the docstring on the function

nathggns replied:
[`5177fad` (#872)](https://github.com/hashintel/hash/pull/872/commits/5177fad45ecc38b16a6d6d04ed0decf7d722dc45)

CiaranMn replied:
Thank you!

CiaranMn commented on packages/hash/shared/src/ProsemirrorManager.ts:
```diff
@@ -0,0 +1,530 @@
+
+type ComponentNodeViewFactory = (meta: HashBlockMeta) => NodeViewFactory;
+
+const isBlockCompatible = (
```
In what circumstances would we try to load incompatible blocks?

nathggns replied:
This isn't about loading an incompatible block. It's about whether we're being asked to render a block entity with a componentId where that component may not be compatible with the underlying data (i.e, the component ids don't match and also aren't compatible in terms of entity properties – i.e, aren't HASH text blocks). 

I'm going to look at replacing this because I can't why this would happen anymore. I think this is legacy logic from a previous iteration of block switching.

nathggns replied:
[`f80e4ec` (#872)](https://github.com/hashintel/hash/pull/872/commits/f80e4ec6c2dbf7d766f91181067704c192ced59c)

CiaranMn replied:
Makes sense, thank you

teenoh commented on packages/hash/shared/src/blocks.ts:
```diff
@@ -217,26 +214,47 @@ export const fetchBlockMeta = async (
+  "https://blockprotocol.org/blocks/@hash/video",
+];
+
+export const isTextBlock = (componentId: string = "") =>
```
I think we should add a comment saying this only checks against a set of defined blocks. This will return false for any text block that isn't in `textBlocks`

nathggns replied:
I've added some comments, but also done on better. I realised in all but one place, this was only being used to check block compatibility, so I've built a better abstraction which is more explicit about this. I've also removed the other use because I realised it wasn't necessary. So we've ended up in what should be a much more readable position. It also fixes a bug where block switching is unexpectedly enabled when there's an error inside the block. [`a721d7a` (#872)](https://github.com/hashintel/hash/pull/872/commits/a721d7a92d7762568bb05b01f644d7b3387bc39a)

CiaranMn commented on packages/hash/shared/src/ProsemirrorManager.ts:
```diff
@@ -0,0 +1,530 @@
+    this.view.dispatch(tr);
+  }
+
+  private async createNewDraftBlock(
```
Maybe we could briefly explain what `draft` is for here? I don't think we do anywhere really

nathggns replied:
I actually ended up renaming this in [`b663a10` (#872)](https://github.com/hashintel/hash/pull/872/commits/b663a10be8b24adc6c08577c1274f2f4d605ff68) to make it clearer what it does. Not sure this comment is relevant anymore – can you check?

CiaranMn replied:
Yes, agreed, thank you

CiaranMn commented on packages/hash/shared/src/entityStorePlugin.ts:
```diff
@@ -368,25 +361,30 @@ const updateEntityStoreListeners = collect<
+        listeners.delete(listener);
+      }
+    } else {
+      if (!listeners) {
+        listeners = new Set();
+      }
```
```suggestion
      listeners ??= new Set();
```

CiaranMn commented on packages/hash/shared/src/prosemirror.ts:
```diff
@@ -105,6 +19,203 @@ export const isEntityNode = (
+          return [
+            "div",
+            {
+              // @todo this isn't applied because of the node view
```
Are you going to address this in a related / soon task? If not can we expand on it, either here or in a task which we link? It's a bit cryptic at the moment

nathggns replied:
I think `@todo` is misleading here, it's more of a note. My bad. Updated the comment. [`2ce9352` (#872)](https://github.com/hashintel/hash/pull/872/commits/2ce9352e6f0a73d91a2963f3e458a09fba702f03)

CiaranMn replied:
Understand it now, thank you

CiaranMn commented on packages/hash/shared/src/prosemirror.ts:
```diff
@@ -105,6 +19,203 @@ export const isEntityNode = (
+        },
+        parseDOM: [
+          {
+            // @todo is this necessary
+            tag: 'div[data-hash-type="block"]',
```
Similar comment to just above – if this is not something you'll be addressing in a follow up, can we add a bit more context?

nathggns replied:
Updated the comment [`2ce9352` (#872)](https://github.com/hashintel/hash/pull/872/commits/2ce9352e6f0a73d91a2963f3e458a09fba702f03)

CiaranMn commented on packages/hash/shared/src/prosemirror.ts:
```diff
@@ -105,6 +19,203 @@ export const isEntityNode = (
+       * This is serialized as a new line in `createEditorView` when copying to
+       * plain text
+       *
+       * @todo figure out out to do this here
```
A typo here

nathggns replied:
[`c92eceb` (#872)](https://github.com/hashintel/hash/pull/872/commits/c92ecebfcdca33cc21c5d7cd720c7e3dd14785d6)

CiaranMn commented on packages/hash/shared/src/prosemirror.ts:
```diff
@@ -148,3 +259,65 @@ export const findComponentNode = (
+  content: (string | T)[];
+}
+
+export const mutateSchema = (
```
Can you explain what this is doing in a comment please?

nathggns replied:
https://github.com/hashintel/hash/pull/872#:~:text=c92eceb-,Explain%20mutateSchema,-38cd5a7

nathggns replied:
[`38cd5a7` (#872)](https://github.com/hashintel/hash/pull/872/commits/38cd5a7a13a94d48e3459218d6733485d9babe76)

CiaranMn commented:
Things identified in testing:
1. "position out of range" 

This error can occur when switching between pages – the most consistent way I've found to reproduce (but not 100%) is to go to the 'my awesome page' dummy page, go to the middle of the formatted para, hit enter to break it up, then click  'new page' - it will often crash when switching to the new page. Assume something to do with retaining the position that no longer exists on the new page. It also seems to happen when leaving a page to go to a non-page (e.g. account home).
<img width="956" alt="Screenshot 2022-08-01 at 16 38 40" src="https://user-images.githubusercontent.com/37743469/182186759-eb541595-4f36-476d-bff0-99ccc7016092.png">

2. "Draft entity already exists" / "invariant: unexpected structure"

If you have multiple clients viewing the same page, you can recreate this error by creating and deleting paras a few times, e.g. get in a para, hit enter, hit backspace, and loop that for a little bit (not sure what exactly triggers it). This is the same whether you have the same page open in the same session from the same user, or whether you're logged in as separate users in different sessions. Sometimes it's the user/window that's doing stuff that crashes, sometimes it it's the other one.

CiaranMn commented:
3. If you attempt to swap from an existing mock data para/header to a para/header, the text is lost. I assume this is because the mock data uses the new system, but it is being swapped with an assumption of the legacy link system (and `__linkedData` doesn't exist on the original mock data paras). I suggest we don't attempt to do anything more to support swapping to `__linkedData` text paras, and just deal with swapping when removing legacy links.

teenoh commented on packages/hash/shared/src/ProsemirrorManager.ts:
```diff
@@ -0,0 +1,530 @@
+    }
+  }
+
+  async loadPage(currentState: EditorState<Schema>, entities: BlockEntity[]) {
```
I think we should add a comment explaining what this method does

nathggns replied:
[`481f73e` (#872)](https://github.com/hashintel/hash/pull/872/commits/481f73e4a2a7b5655653250c703c7f0f3361689a)

teenoh commented on packages/hash/shared/src/ProsemirrorManager.ts:
```diff
@@ -0,0 +1,530 @@
+      entityStorePluginState(currentState).store.draft,
+    );
+
+    const newNodes = await Promise.all(
```
Perhaps we should just call it `nodes`. `newNodes` makes it feel like it's only new blocks that are defined here, whereas it's actually all blocks in the page

nathggns replied:
[`a59d29a` (#872)](https://github.com/hashintel/hash/pull/872/commits/a59d29af479fb27d66fc480bc6f8fd8ea2cfd771)

teenoh commented on packages/hash/shared/src/ProsemirrorManager.ts:
```diff
@@ -0,0 +1,530 @@
+      ? entityStoreState.store.draft[draftBlockId]
+      : null;
+
+    if (
+      (draftBlockId && !blockEntity) ||
+      (blockEntity && !isDraftBlockEntity(blockEntity))
+    ) {
+      throw new Error("draft id does not belong to a block");
+    }
```
I feel breaking this into 2 conditionals will make errors here easier to debug

```suggestion
   if (draftBlockId && !blockEntity){
    throw new Error("draft id is not present in entity store")
   }
   
    if (blockEntity && !isDraftBlockEntity(blockEntity)) {
      throw new Error("draft id does not belong to a block");
    }
```

nathggns replied:
I think there's a formatting issue with the suggestion, so did this manually, but the suggestion has been accepted in [`d382961` (#872)](https://github.com/hashintel/hash/pull/872/commits/d382961d9f1ea3e7b41b532cdc38411851d9cf39)

teenoh commented on packages/hash/shared/src/ProsemirrorManager.ts:
```diff
@@ -0,0 +1,530 @@
+             */
+            {
+              ...entityProperties,
+              text: blockEntity.properties.entity.properties.text,
```
Sorry, I might have missed this somewhere. I thought the structure of text block properties looked more like

```
blockEntity.properties.entity.properties.tokens
```

Are we replacing `tokens` with `text` now?

nathggns replied:
No. `text` is the property on which legacy links are store *to* text entities *from* block data entities (which may store a heading level, for example). Blocks shouldn't link directly to text entities. There's no change here.

teenoh commented on packages/hash/shared/src/ProsemirrorManager.ts:
```diff
@@ -0,0 +1,530 @@
+    view.dispatch(tr);
+  }
+
+  async createBlock(
```
It feels this method also handles updating the block/swapping. If we can't rename, then we should probably add a comment explaining what it does

nathggns replied:
This never updates a block itself. It can update block data.

The purpose of this block is to render a component specified by a `targetComponentId` with a block specified by a `draftBlockId`. It may update the block data if a specific variant of the component is specified.

Mostly it just works out if the `draftBlockId` and `targetComponentId` are compatible, and creates a new block if not.

teenoh reviewed (commented):
Thanks for this @nathggns! I know this was definitely not a walk in the park. 
This simplifies/organizes things a lot but I think we should also use the opportunity to add comments like @CiaranMn mentioned, so it's easier to follow through the code

teenoh commented:
I also get the "Draft entity already exists" / "invariant: unexpected structure" error @CiaranMn pointed out. It can be reproduced by changing the entity type of the table block in `Table Testing page`

nathggns commented:
> 3. If you attempt to swap from an existing mock data para/header to a para/header, the text is lost. I assume this is because the mock data uses the new system, but it is being swapped with an assumption of the legacy link system (and `__linkedData` doesn't exist on the original mock data paras). I suggest we don't attempt to do anything more to support swapping to `__linkedData` text paras, and just deal with swapping when removing legacy links.

@CiaranMn this is actually a bug i've introduced so decided to fix it, as Joseph needs to be able to use the product over the next week+ while I'm away. Fixed in [`d5f8edf` (#872)](https://github.com/hashintel/hash/pull/872/commits/d5f8edffc3d855392c2dbf7b45bfc7c1b1b26af5)

nathggns commented:
@CiaranMn @teenoh I've fixed both the "position out of range" bug [`0ea88e6` (#872)](https://github.com/hashintel/hash/pull/872/commits/0ea88e606c344708ec5f1c3919493cbd9b4a7be8) and also the "draft entity already exists" bug [`eb0e5b0` (#872)](https://github.com/hashintel/hash/pull/872/commits/eb0e5b0a5f479ea744ad376c0dce300af359ed75)

nathggns commented:
> I get this error when changing the entity type of the table block in `Table Testing page`

@teenoh I can't reproduce. Can you check again after my recent changes, or record with replay.io? It's possible my bug fix at [`eb0e5b0` (#872)](https://github.com/hashintel/hash/pull/872/commits/eb0e5b0a5f479ea744ad376c0dce300af359ed75) fixed it.

CiaranMn commented:
I'm unable to recreate 'position out of range', but I can still recreate 'draft entity already exists'

https://user-images.githubusercontent.com/37743469/182381332-47b1bd56-97f6-4303-8042-0fbfff86c1e4.mp4

teenoh commented:
> > I get this error when changing the entity type of the table block in `Table Testing page`
> 
> @teenoh I can't reproduce. Can you check again after my recent changes, or record with replay.io? It's possible my bug fix at [`eb0e5b0` (#872)](https://github.com/hashintel/hash/pull/872/commits/eb0e5b0a5f479ea744ad376c0dce300af359ed75) fixed it.

Thanks @nathggns. I tested with recent changes and no longer get the error

nathggns commented on packages/hash/api/src/collab/Instance.ts:
```diff
@@ -514,12 +514,12 @@ export class Instance {
+         * This is a *serious* security risk as the frontend can instruct us
          * to make a web request
+         *
+         * @todo deal with this?
```
Made a new ticket tracking this problem.

nathggns commented on packages/hash/api/src/collab/save.ts:
```diff
@@ -126,6 +126,57 @@ const calculateSaveActions = async (
   const draftIdToPlaceholderId = new Map<string, string>();
   const draftIdToBlockEntities = new Map<string, DraftEntity<BlockEntity>>();
 
+  const shouldSetTextProperties = <T extends {}>(
```
I've renamed this to `ensureNecessaryLegacyTextLink` to make it clearer what's going on here, and added a comment. I've also made its return type more sensible. I can't move it to outside of the function because it mutates something in local scope. I could make it an argument but I think that makes it even less clear what is going on.

nathggns commented on packages/hash/api/src/collab/save.ts:
```diff
@@ -126,6 +126,57 @@ const calculateSaveActions = async (
+      if (
+        !isEqual(savedWithoutText, draftWithoutText) ||
+        isEqual(savedTextLink, properties.text.__linkedData)
+      ) {
```
I think this might be a bug, but I'm not actually certain. It's a replication of the same logic in main. The idea here is we don't want to update an entity just because the text entity within it has changed (as that is processed separately), but I think I got it wrong originally and therefore in this refactor.

I've rewritten this whole process in [`b807520` (#872)](https://github.com/hashintel/hash/pull/872/commits/b807520cb738d9b1d0c6b25adcb63fb5d50ceb2f) to make this significantly clearer

CiaranMn replied:
Looks good, thank you

nathggns commented on packages/hash/api/src/collab/save.ts:
```diff
@@ -146,45 +197,33 @@ const calculateSaveActions = async (
-        )
-          ? savedEntity.properties.text.__linkedData
-          : null;
+      const nextProperties = shouldSetTextProperties(
```
This was rewritten in [`b807520` (#872)](https://github.com/hashintel/hash/pull/872/commits/b807520cb738d9b1d0c6b25adcb63fb5d50ceb2f) so the comment is out of date (and I incidentally introduced a `nextProperties` variable here anyway!). 

Could you check if the new logic is sufficiently readable?

CiaranMn replied:
Perfectly clear, thank you

nathggns commented on packages/hash/api/src/collab/save.ts:
```diff
@@ -146,45 +197,33 @@ const calculateSaveActions = async (
     } else {
-      const placeholderId = generatePlaceholderId();
-      draftIdToPlaceholderId.set(draftEntity.draftId, placeholderId);
+      const dependedOn = draftIdToPlaceholderId.has(draftEntity.draftId);
```
Done! [`6aa5155` (#872)](https://github.com/hashintel/hash/pull/872/commits/6aa5155b8f7fcbc308e1f90d9a0721904f3a95b8)

CiaranMn replied:
Looks good, thank you

nathggns commented on packages/hash/api/src/collab/save.ts:
```diff
@@ -126,6 +126,57 @@ const calculateSaveActions = async (
+    return null;
+  };
+
   for (const draftEntity of Object.values(store.draft)) {
```
Done in [`c9ccff3` (#872)](https://github.com/hashintel/hash/pull/872/commits/c9ccff3f6a9573aa9f3e8f050f47a631c4745b56)

CiaranMn replied:
Excellent, thanks!

nathggns commented on packages/hash/api/src/collab/save.ts:
```diff
@@ -245,12 +265,19 @@ const calculateSaveActions = async (
-      });
+      };
+
+      if (dependedOn) {
+        const idx = actions.findIndex((item) => !item.createEntityType);
+        actions.splice(idx === -1 ? 0 : idx, 0, action);
+      } else {
+        actions.push(action);
```
I think I've inserted my own comment which helps explain this – let me know if you think this is sufficient

CiaranMn replied:
Even better, thank you

nathggns commented on packages/hash/frontend/src/blocks/page/ComponentView.tsx:
```diff
@@ -66,66 +76,64 @@ export const componentViewTargetSelector = "div[data-target=true]";
       this.update(this.node);
     });
 
+    this.wasSuggested =
+      suggesterPluginKey.getState(this.editorView.state)
+        ?.suggestedBlockPosition === this.getPos();
+
+    if (this.wasSuggested) {
+      this.editorView.dispatch(
+        this.editorView.state.tr.setMeta(suggesterPluginKey, {
+          type: "suggestedBlock",
+          payload: { position: null },
+        } as SuggesterAction),
+      );
+    }
```
[`1bd4dba` (#872)](https://github.com/hashintel/hash/pull/872/commits/1bd4dba2e7321062a2dee6cc11e1df637c6da04e)

CiaranMn replied:
Perfect - this is the sort of comment I think we should aim to sprinkle everywhere in the PM code as we refactor it

nathggns commented on packages/hash/frontend/src/blocks/page/ComponentView.tsx:
```diff
@@ -66,66 +76,64 @@ export const componentViewTargetSelector = "div[data-target=true]";
-      const blockDraftId: string = this.editorView.state.doc
-        .resolve(this.getPos())
-        .node(2).attrs.draftId;
+    if (
+      isComponentNode(node) &&
+      componentNodeToId(node) === this.meta.componentId
+    ) {
```
Prosemirror will sometimes call `update` on your `NodeView` with new nodes to see if they're compatible, so that they can be reused. If you return `false` from the `update` function, it will call `destroy` on your `NodeView` and create a new one instead. So this means in theory we could get `update` called with a component node representing a different component. This isn't supported, so we return false. See https://prosemirror.net/docs/ref/#view.NodeView.update

This is only something I know Prosemirror might do in theory, but haven't tested it. This logic is already there in main though - https://github.com/hashintel/hash/blob/main/packages/hash/frontend/src/blocks/page/ComponentView.tsx#L123

I'll add a comment explaining this.

nathggns replied:
[`d28d807` (#872)](https://github.com/hashintel/hash/pull/872/commits/d28d80776defe42df585b607479ec9bb44d173b5)

CiaranMn replied:
Understood, thank you

CiaranMn commented:
Thanks for taking the time to add the comments and other changes @nathggns – I think they've significantly improved any reader's understanding of the code. All my comments except [this crash](https://github.com/hashintel/hash/pull/872#issuecomment-1202503031) have been addressed

nathggns commented:
@CiaranMn I've added a few more commits to aim to improve stability here, and think I found why you're still getting the bug. Can you try again?

[`787e01f` (#872)](https://github.com/hashintel/hash/pull/872/commits/787e01f79f986e86bb12c00020df375f9a7f8c9b)
[`7d59ef0` (#872)](https://github.com/hashintel/hash/pull/872/commits/7d59ef0c6f552fbb99e3803e971d23ae6308b189)
[`8478fb9` (#872)](https://github.com/hashintel/hash/pull/872/commits/8478fb90d962795f99bd86a82b441d8a4ca5aa94)

CiaranMn reviewed (approved):
I can't recreate the previous crashes, looks good to me – great work @nathggns! Some very helpful refactoring and comments introduced here, as well as the groundwork for the hook service
